xarray/namedarray/core.py

```diff
@@ -0,0 +1,434 @@
+    # ):
+    #     return typing.cast(T_DuckArray, data)
+    if isinstance(data, tuple):
+        data = to_0d_object_array(data)
+
+    # validate whether the data is valid data types.
+    return typing.cast(T_DuckArray, np.asarray(data))
+
+
+class NamedArray:
+    __slots__ = ("_dims", "_data", "_attrs")
+
+    def __init__(
+        self,
+        dims: DimsInput,
+        data: T_DuckArray | np.typing.ArrayLike,
+        attrs: dict | None = None,
+    ):
+        self._data: T_DuckArray = as_compatible_data(data)
+        self._dims: Dims = self._parse_dimensions(dims)
+        self._attrs: dict | None = dict(attrs) if attrs else None
+
+    @property
+    def ndim(self: T_NamedArray) -> int:
+        """
+        Number of array dimensions.
+
+        See Also
+        --------
+        numpy.ndarray.ndim
+        """
+        return len(self.shape)
+
+    @property
+    def size(self: T_NamedArray) -> int:
+        """
+        Number of elements in the array.
+
+        Equal to ``np.prod(a.shape)``, i.e., the product of the array’s dimensions.
+
+        See Also
+        --------
+        numpy.ndarray.size
+        """
+        return math.prod(self.shape)
+
+    def __len__(self) -> int:
+        try:
+            return self.shape[0]
+        except Exception as exc:
+            raise TypeError("len() of unsized object") from exc
+
+    @property
+    def dtype(self: T_NamedArray) -> np.dtype:
+        """
+        Data-type of the array’s elements.
+
+        See Also
+        --------
+        ndarray.dtype
+        numpy.dtype
+        """
+        return self._data.dtype
+
+    @property
+    def shape(self: T_NamedArray) -> tuple[int, ...]:
+        """
+        Tuple of array dimensions.
+
+        See Also
+        --------
+        numpy.ndarray.shape
+        """
+        return self._data.shape
+
+    @property
+    def nbytes(self: T_NamedArray) -> int:
+        """
+        Total bytes consumed by the elements of the data array.
+
+        If the underlying data array does not include ``nbytes``, estimates
+        the bytes consumed based on the ``size`` and ``dtype``.
+        """
+        if hasattr(self._data, "nbytes"):
+            return self._data.nbytes
+        else:
+            return self.size * self.dtype.itemsize
+
+    @property
+    def dims(self: T_NamedArray) -> Dims:
+        """Tuple of dimension names with which this variable is associated."""
+        return self._dims
+
+    @dims.setter
+    def dims(self: T_NamedArray, value: DimsInput) -> None:
+        self._dims = self._parse_dimensions(value)
+
+    def _parse_dimensions(self: T_NamedArray, dims: DimsInput) -> Dims:
+        dims = (dims,) if isinstance(dims, str) else tuple(dims)
+        if len(dims) != self.ndim:
+            raise ValueError(
+                f"dimensions {dims} must have the same length as the "
+                f"number of data dimensions, ndim={self.ndim}"
+            )
+        return dims
```

Comment:
This allows `dims = (None, )`. Is that ok?
The problem is that we often use `None` as the default value for functions having a dim argument.

Reply:
> This allows dims = (None, ). Is that ok? 

it is my understanding that this particular situation of `dims=(None, )` never occurs since `tuple(None)` doesn't work, right?

Reply:
I see... thank you for clarifying. Curious to hear your thoughts on how we should handle this

Reply:
`None` is not really practical in current xarray so not allowing it as a dimension is probably the easiest path, but type hinting will not be correct.

I want `dims` to have a type hint that is consistent, easy to read and understand. In a dream world it would look something like this:
```python 
InputDim = Hashable # Single dimension
InputDims = Iterable[InputDim , ...] # multiple dimensions
InputDimOrDims = Union[InputDim, InputDims]  # Single or multiple dimensions
```

Then we can easily go through our xarray methods and easily replace `dim` and `dims` arguments.

`Hashable` could be fine in `NamedArray`, we haven't introduced `None` as a typical default value there yet.

But it wouldn't be easy in xarray because we use `None` as default value a lot, which will (I suspect) lead to a bunch of refactoring and deprecations. I haven't tried it maybe it's doable?

Another idea is to try and make a HashableExcludingNone:
```python
HashableExcludingNone = Union[int, str, tuple, ...] # How many more Hashables are there?
InputDim = HashableExcludingNone # Single dimension
InputDims = Iterable[InputDim , ...] # multiple dimensions
InputDimOrDims = Union[InputDim, InputDims]  # Single or multiple dimensions
```
I suspect this is harder than it seems.

Another idea is drop the idea of Hashable and just allow a few common ones that are used:
```python 
InputDim = str # Single dimension
InputDims = tuple[InputDim , ...] # multiple dimensions
InputDimOrDims = Union[InputDim, InputDims]  # Single or multiple dimensions
```
Very clear! I think a few users (and maintainers) will be sad because of the lack of flexibility though.

No easy paths, and trying to be backwards compatible is very demotivating.